internal/server/application.go

```diff
@@ -0,0 +1,169 @@
+package server
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/corazawaf/coraza-spoa/internal/cache"
+	"github.com/corazawaf/coraza-spoa/internal/config"
+	"github.com/corazawaf/coraza-spoa/internal/logger"
+	"github.com/corazawaf/coraza/v3"
+	"github.com/corazawaf/coraza/v3/types"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+type ErrInterrupted struct {
+	Interruption *types.Interruption
+}
+
+func (e ErrInterrupted) Error() string {
+	return fmt.Sprintf("interrupted with status %d and action %s", e.Interruption.Status, e.Interruption.Action)
+}
+
+func (e *ErrInterrupted) Is(target error) bool {
+	t, ok := target.(*ErrInterrupted)
+	if !ok {
+		return false
+	}
+	return e.Interruption == t.Interruption
+}
+
+type application struct {
+	config.Application
+	waf    coraza.WAF
+	logger *zerolog.Logger
+}
+
+func (a *application) HandleRequest(req *applicationRequest) error {
+	id := req.ID
+	if id == "" {
+		return fmt.Errorf("request id is empty")
+	}
+	tx := a.waf.NewTransactionWithID(id)
+	tx.ProcessConnection(req.SrcIp.String(), int(req.SrcPort), req.DstIp.String(), int(req.DstPort))
+	url := strings.Builder{}
+	url.WriteString(req.Path)
+	if req.Query != "" {
+		url.Grow(len(req.Query) + 1)
```

Comment:
Maybe use the url.URL structure? It also contains construction based on RawQuery inputs